src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts

```diff
@@ -683,4 +692,146 @@ export class InstructorCourseEditPageComponent implements OnInit {
+            return;
+          }
+
+          this.addNewInstructors(instructors, modalRef);
+        });
+      });
+    });
+  }
+
+  /**
+   * Adds new instructors.
+   */
+  addNewInstructors(instructors: Instructor[], modalRef: NgbModalRef): void {
+    of(...instructors).pipe(
+      concatMap((instructor: Instructor) => {
+        return this.instructorService.createInstructor({
+          courseId: this.courseId,
+          requestBody: {
+            name: instructor.name,
+            email: instructor.email,
+            role: instructor.role!,
+            displayName: instructor.displayedToStudentsAs,
+            isDisplayedToStudent: instructor.isDisplayedToStudents!,
+          },
+        });
+      }),
+      // always close the modal after it enters the last step no matter adding succeeds or fails
+      finalize(() => {
+        this.isCopyingInstructor = false;
+        modalRef.componentInstance.isCopyingSelectedInstructors = false;
+        modalRef.close();
+      }),
+    ).subscribe((newInstructor: Instructor) => {
+      const newDetailPanels: InstructorEditPanelDetail = {
+        originalInstructor: { ...newInstructor },
+        originalPanel: this.getInstructorEditPanelModel(newInstructor),
+        editPanel: this.getInstructorEditPanelModel(newInstructor),
+      };
+      newDetailPanels.editPanel.permission = this.newInstructorPanel.permission;
+      newDetailPanels.originalPanel = JSON.parse(JSON.stringify(newDetailPanels.editPanel));
+
+      this.instructorDetailPanels.push(newDetailPanels);
+    }, (err: ErrorMessageOutput) => {
+      this.statusMessageService.showErrorToast(err.error.message);
+    }, () => {
+      this.statusMessageService.showSuccessToast(`The selected instructor(s) have been added successfully.
+      An email containing how to 'join' this course will be sent to them in a few minutes.`);
+    });
+  }
+
+  /**
+   * Verifies that no two selected instructors have the same email addresses and any selected instructor's
+   * email addresses already exists in the course.
+   */
+  verifyInstructorsToCopy(instructors: Instructor[]): Observable<boolean> {
+    return forkJoin([
+      this.instructorService.loadInstructors({
+        courseId: this.courseId,
+        intent: Intent.FULL_DETAIL,
+      }),
+    ]).pipe(
+      map((values: [Instructors]) => this.hasNoRepeatedEmail(instructors.concat(values[0].instructors))),
+    );
+  }
+
+  /**
+   * Checks if the specified array contains two instructors with the same email address.
+   */
+  hasNoRepeatedEmail(instructors: Instructor[]): boolean {
+    const emailSet: Set<string> = new Set();
+    for (const instructor of instructors) {
+      if (emailSet.has(instructor.email)) {
+        this.statusMessageService.showErrorToast(`An instructor with email address ${instructor.email} 
+        already exists in the course and/or you have selected more than one instructor with this email address.`);
```

Comment:
The method and the header makes no mention that an error toast is involved.

Reply:
Your method name is `hasNoRepeatedEmail`. One look at it (and the header) suggests that it does nothing but a check and a boolean return. If there is an unexpected behavior in the method body that requires actually reading the code, the implementation is questionable.

Either:
- Don't put the error toast here. The error toast should be at the call site.
- If the error toast absolutely needs to be in here, the method should be renamed/reexplained.

Reply:
I will simply delete the helper ethod and do everything in its caller instead. The method seems redundant to me.